read-excel/storage.hpp

```diff
@@ -0,0 +1,36 @@
+#pragma once
+// C++ include.
+#include <string>
+
+// Excel include.
+#include "formula.hpp"
+
+namespace Excel {
+
+//
+// IStorage
+//
+
+//! Excel storage interface.
+struct IStorage {
+	~IStorage() {}
+	virtual void onSharedString( size_t sstSize, size_t idx, const std::wstring & value ) = 0;
+	virtual void onDateMode( uint16_t mode ) = 0;
+	virtual void onSheet( size_t idx, const std::wstring & value ) = 0;
+	virtual void onCellSharedString( size_t sheetIdx, size_t row, size_t column, size_t sstIndex ) = 0;
+	virtual void onCell( size_t sheetIdx, size_t row, size_t column, const std::wstring & value ) = 0;
+	virtual void onCell( size_t sheetIdx, size_t row, size_t column, double value ) = 0;
+	virtual void onCell( size_t sheetIdx, const Formula & value ) = 0;
+}; // struct IStorage

```

Comment:
To avoid mess in `Excel::Book` public API, please, do the following.

```suggestion
struct IStorage;

class XLSParser final {
public:
	static void loadBook( ..., IStorage & storage );
}; // class XLSParser

struct IStorage {
	virtual ~IStorage() {}
	
protected:
	friend class XLSParser;
	
	virtual void onSharedString( size_t sstSize, size_t idx, const std::wstring & value ) = 0;
	virtual void onDateMode( uint16_t mode ) = 0;
	virtual void onSheet( size_t idx, const std::wstring & value ) = 0;
	virtual void onCellSharedString( size_t sheetIdx, size_t row, size_t column, size_t sstIndex ) = 0;
	virtual void onCell( size_t sheetIdx, size_t row, size_t column, const std::wstring & value ) = 0;
	virtual void onCell( size_t sheetIdx, size_t row, size_t column, double value ) = 0;
	virtual void onCell( size_t sheetIdx, const Formula & value ) = 0;
}; // struct IStorage
```

Reply:
Do you suggest to move loadBook into storage.hpp? No problem, but my main point was holding project history clear and understandable. That's why I leave all functions at their places, it makes possible to review all changes in easy way.
Now you can see, what exactly was changed in loadBook function, after moving it would be more complicated.

Reply:
Why do you need this 'friend class XLSParser'? There are 11 functions which deal with IStorage, should they be static members of XLSParser?

Reply:
Not mandatory to move `loadBook` into  `storage.hpp`. I just showed an idea. Better to move `XLSParser` with all that 11 functions to separate header, to store parsing functions in one place.

Yes, that parsing functions should be static members of `XLSParser`.

The main - `on....()` methods of `IStorage` should be protected, and in all derivative classes.

Reply:
Sorry for the late answer, didn't see notification.

Reply:
ok, i see - you hate free functions ))) will do.